File: SchoolData/src/main/java/org/gb/homework/schooldata/service/impls/UserServiceImpl.java
package org.gb.homework.schooldata.service.impls;

import org.gb.homework.schooldata.model.User;
import org.gb.homework.schooldata.repository.Repository;
import org.gb.homework.schooldata.service.UserService;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;

/**
 *
 */
public class UserServiceImpl<T extends User> implements UserService<T> {
    protected final Repository<T> repository;

    public UserServiceImpl(Repository<T> repository) {
        this.repository = repository;
    }

    @Override
    public List<T> getAllUsers() {
        return repository.getAll().stream().filter(User::isActual).toList();
    }

    @Override
    public void add(T user) {
        repository.save(user);

    }

    @Override
    public void remove(T user) {
        repository.remove(user);
    }

    @Override
    public int getNextId() {
        Optional<T> last = repository.getAll().stream().max(Comparator.comparingInt(User::getId));

        return last.isPresent() ? last.get().getId() + 1 : 0;
    }

    @Override
    public void edit(T user) {
        repository.save(user);
    }
}

File: csharp/SchoolData/Service/Impls/UserServiceImpl.cs
using Gb.Homework.SchoolData.Model;
using Gb.Homework.SchoolData.Repositories;

namespace Gb.Homework.SchoolData.Service.Impls
{

    public class UserServiceImpl<T> : ISchoolUserService<T> where T : User
    {
        protected Repository<T> repository;

        public UserServiceImpl(Repository<T> repository)
        {
            this.repository = repository;
        }

        public List<T> GetAllUsers()
        {
            return repository.All.Where(x => x.IsActual).ToList();
        }

        public void Add(T user)
        {
            repository.Save(user);
        }

        public void Edit(T user)
        {
            repository.Save(user);
        }

        public void Remove(T user)
        {
            repository.Remove(user);
        }

        public int GetNextId()
        {
            var last = repository.All.MaxBy(x => x.Id);

            return last != null ? last.Id + 1 : 0;
        }

        List<User> IUserService.GetAllUsers()
        {
            return GetAllUsers().Cast<User>().ToList();
        }

        void IUserService.Add(User user)
        {
            Add((T)user);
        }

        void IUserService.Edit(User user)
        {
            Edit((T)user);
        }

        void IUserService.Remove(User user)
        {
            Remove((T)user);
        }
    }

}